### ISSUE TITLE
Distribution of Modelica libraries - Modelica Archive proposal

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1152]
**Reported by dhedberg on 31 May 2013 12:35 UTC**
This is a proposal on how Modelica libraries could be distributed in a standardized single-file Modelica archive, and how this archive file includes enough information for Modelica tools and web services (providing downloads of archive files) to obtain general information on the library, resolve dependences on other libraries, automate the installation of the library, etc., without parsing the contained Modelica files.

See attached document.

Related tickets: #536

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1152

**Comment by modelica-trac-importer** [comment]
**Modified by dhedberg on 31 May 2013 12:39 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 31 May 2013 12:51 UTC**
So annoying to read doc-files...
7. Why not use the naming scheme for .mo-files also for the .mar? (Using spaces rather than the dash).
Note that you do have libraries that contain libraries of different version (like Modelica 3.2, Complex 1.0, etc). So you would likely need to read the manifest regardless when it comes to resolving uses-annotations of libraries not using the archive format. You stated there are no backwards compatibility issues. This is one.
11. Why specify PNG? Modelica icons are vector graphics, so surely SVG would be more suitable?

Regarding companion libraries, would it not suffice to add recursion to format? Or simply to ship a zip of .mar's.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 1 Jun 2013 08:51 UTC**
The proposal appears to have a lot of overlap with already existing things. See Modelica 3.3 spec:
- section 18.2 Annotations for Documentation
- section 18.8 Annotations for Version Handling
- section 12.9.4 Annotations for External Libraries and Include Files
- section 13.2.2 Mapping Package/Class Structures to a Hierarchical File System
besides more. 

Requirement 4 appears questionable _"The manifest file should be in the XML format as this format is supported by all web services and used by FMI"_. FMI goes beyond Modelica, this is why lacking a common syntax and naturally selected XML in 2010. Web services are increasingly moving away from XML towards JSON: The Fat-Free Alternative to XML. See http://www.json.org/xml.html or in one sentence from a Modelican's point of view: JSON is a sub-set of JavaScript, comparable to parameter records in Modelica. So why not using parameter records and annotations for the distribution of Modelica libraries? 

In other words: if you want to distribute a Modelica library in one file, why not just zip the corresponding directory in the file system? What is missing in Modelica 3.3?

**Comment by modelica-trac-importer** [comment]
**Comment by dhedberg on 2 Jun 2013 15:39 UTC**
sjoelund.se:

We could definitely use a space instead of a dash.

The idea is that Commplex 1.0 would be included in the Modelica-3.2.mar as a companion library and hence no dependency of Complex would be listed in the manifest as it would be resolved internally. If there are libraries distributed in any other way than an archive file, then the tool would have no other way then to rely on the uses annotation, but I don't see any backwards incompatibly issues here?

SVG would be an alternative/complement to PNG, but it would also increase the implementation effort as Modelica tools would have to implement a SVG generator.

There is nothing that stops you from distributing the MSL in separate archive files for each companion library with the current proposal (Complex, ObsoleteModelica3, ModelicaServices, etc). However, it also lets you distribute MSL in one single archive file if you would like to do so.

**Comment by modelica-trac-importer** [comment]
**Comment by dhedberg on 2 Jun 2013 15:46 UTC**
rfranke:

JSON is a valid alternative to XML and I agree that it is more lightweight than XML and if more people prefer JSON over XML I am absolutely fine with that.

Regarding the overlap, yes, there is an overlap, but the manifest file is there to make it possible for non-Modelica tools, such as web services to easily present the details/contents of an archive file and to resolve dependencies without knowledge of Modelica (integration of a Modelica parser).

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 3 Jun 2013 09:07 UTC**
What I would have liked is a structure like:
```xml
<library name="...">
  ...
  <compatibility>...</compatibility>
  <dependencies>...</dependencies>
</library>
<library>
  ...
</library>
<library>
  ...
</library>
```

Rather than companion libraries. This feels simpler to me anyway. Which is the main library would be tagged by the loadByDefault anyway.

I would also have liked to see all the libraries in the top-level, so:
```
MyFile.mar/Modelica 3.2.1/package.mo
MyFile.mar/ModelicaServices 3.2.1/package.mo
MyFile.mar/ModelicaReference 3.2.1/package.mo
```
(This so tools have a simple solution: unzip the archive to a single directory without any intelligence)

**Comment by modelica-trac-importer** [comment]
**Comment by fcasella on 3 Jun 2013 09:15 UTC**
Please also consider the issue of external C sources and binary libraries (see discussion in #1155)

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 3 Jun 2013 10:29 UTC**
Replying to [comment:3 rfranke]:
> The proposal appears to have a lot of overlap with already existing things. See Modelica 3.3 spec:

...
> In other words: if you want to distribute a Modelica library in one file, why not just zip the corresponding directory in the file system? What is missing in Modelica 3.3?

Written on the plane before seeing the other texts:
Just zipping it seems like an excellent idea, and to answer the question about missing features in 3.3 I can only think of one:
Bundled packages used in implementation; as a cleaner and safer way to handle versions of packages.

There are still some unclear choices in the design - but let me present two options below - with some questions:

As an example assume you want to make a library ElectricMotor and you find that you can build its components using the libraries Electrical and Rotational (not parts of MSL).

The goals would be:
* Developer of ElectricMotor can use Electrical and Rotational.
* Users of ElectricMotor don't want to care about Electrical and Rotational.
* Both want ElectricMotor to use Electrical as a tested version.

1. The problem is then to distribute ElectricMotor-library and the idea would be:
```
ElectricMotor
   package.mo
   Protected // Or other name
      Electrical 4.5.mo
      Rotational 2.3
        package.mo
        ...
```
The one getting the library will clearly see that "ElectricMotor" is the package they intend to use; and also that this exact configuration is fully tested.

From the perspective of classes inside ElectricMotor the "Protected" packages work exactly like global packages with those names; and the only downside is that users might waste disk-space if there are multiple packages like ElectricMotor using Electrical. However, disk-space is cheap and getting cheaper - and it might not even be an issue.

The design questions are:
* This design would allow multiple packages to include different versions of Electrical. Thus versioned lookup will be local to each "package-bundle" such that everything in ElectricMotor-bundle use consistent versions.
* Should it be possible to access the Electrical-package used in ElectricMotor in some way from the outside? Yes/No/Up to designer of ElectrialMotor? E.g. ElectricalMotor.Public.Electrical?
* Why is it necessary to include Electrical 4.5 if we know how to find it?

2. Extended uses-annotations to allow them to give URLs (or URIs) as well.
If packages can be distributed as zip-files we can give them stable URLs; and the version-annotation could refer to that:
So ElectricMotor's package.mo could contain:
  annotations(uses(Electrical(version="4.5", url="http://www.modelica.org/Libraries/Electrical 4.5.zip", Rotational(version="2.3", url="http://www.modelica.org/libraris/Rotational 2.3.zip")));

This could be seen as a completely backward compatible solution where the url is only intended for users (so all the tool has to do is to tell the user that Electrical 4.5 is missing and it can be found at this url ...) or we could combine it with elements of the first idea.

The last part might require some explanation:
Firstly Electrical and Rotational can be seen as "internal" packages and the user will only directly see ElectricMotor as a top-level package. 
Secondly is that we assume that ElectricMotor was tested with Electrical 4.5 and the end-user has installed a redesigned Electrical 5. The ElectricMotor-library can still use Electrical 4.5 - and from a lookup perspective it is as if we add an intermediate level between top-level packages and the top of the package hierarchy. This solves the problem of finding Electrical 5 vs. Electrical 4.5 without having to use "Electrical 4.5.Resistor" or something equally awkward.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 4 Jun 2013 08:27 UTC**
What I still have not been able to figure out in this proposal is why the extra manifest information (e.g. about versions and dependencies) cannot be present in the normal directories for Modelica libraries.

The basic idea would then be that we add an additional "manifest" file (with redundant information) when Modelica packages are stored on a file-system (similarly as package.order it is handled differently for data-base storage).

That information could be used to speed up loading of libraries - independently of whether the library is in archive format or not.
\---
The basic idea is that the Modelica archive is then just the zip of a folder in the MODELICAPATH - and the user can unzip it and place it in the MODELICAPATH. 

This is different from .zip-variants like .fmi, .jar, .docx. The main idea is that users are interested are looking at the contents of a Modelica-archive and there is no need to hide that.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 4 Jun 2013 08:36 UTC**
Replying to [comment:9 hansolsson]:
> That information could be used to speed up loading of libraries - independently of whether the library is in archive format or not.

Or perhaps more important: to tag which version of the grammar the library is written in.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 15 Nov 2013 13:30 UTC**
80th Design Meeting is long over ... closing that milestone and still open tickets set back to \*undecided\*.

**Comment by modelica-trac-importer** [comment]
**Comment by pharman on 5 Jan 2016 10:03 UTC**
See also #132

**Comment by HansOlsson** [comment]
It is still (after 6 years) not clear why the archive proposal wouldn't just zip existing libraries; possibly in combination with additional information in those directories. Thus more info is needed.
(The language-version was one reason, but that has a separate MCP.)
